src/pure_aworset.erl

```diff
@@ -0,0 +1,219 @@
+            1;
+        false ->
+            0
+    end.
+
+%% @doc Removes redundant operations from POLog of `pure_aworset()'
+%% Called upon updating (add, rmv) the `pure_aworset()'
+-spec remove_redundant_polog({pure_type:id(), pure_aworset_op()}, pure_aworset()) -> {boolean(), pure_aworset()}.
+remove_redundant_polog({VV1, Op}, {?TYPE, {POLog0, ORSet}}) ->
+    POLog1 = orddict:fold(
+        fun(Key, Value, Acc) ->
+            case redundant({Key, Value}, {VV1, Op}) of
+                0 ->
+                    orddict:store(Key, Value, Acc);
+                1 ->
+                    Acc
+            end
+        end,
+        orddict:new(),
+        POLog0
+    ),
+    {true, {?TYPE, {POLog1, ORSet}}}.
+
+%% @doc Removes redundant operations from POLog of `pure_aworset()'
+%% Called upon updating (add, rmv) the `pure_aworset()'
+-spec remove_redundant_crystal({pure_type:id(), pure_aworset_op()}, pure_aworset()) -> {boolean(), pure_aworset()}.
+remove_redundant_crystal({_VV1, {_X, Elem}}, {?TYPE, {POLog, AWORSet}}) ->
+    case ordsets:is_element(Elem, AWORSet) of
+        true ->
+            {true, {?TYPE, {POLog, ordsets:del_element(Elem, AWORSet)}}};
+        false ->
+            {false, {?TYPE, {POLog, AWORSet}}}
+    end.
+
+%% @doc Checks stable operations and remove them from POLog of `pure_aworset()'
+-spec check_stability(pure_type:id(), pure_aworset()) -> pure_aworset().
+check_stability(StableVV, {?TYPE, {POLog0, AWORSet0}}) ->
+    {POLog1, AWORSet1} = orddict:fold(
+        fun(Key, {_Op, Elem}=Value, {AccPOLog, AccORSet}) ->
+            case pure_trcb:happened_before(Key, StableVV) of
+                true ->
+                    {AccPOLog, ordsets:add_element(Elem, AccORSet)};
+                false ->
+                    {orddict:store(Key, Value, AccPOLog), AccORSet}
+            end
+        end,
+        {orddict:new(), AWORSet0},
+        POLog0
+    ),
+    {?TYPE, {POLog1, AWORSet1}}.
+
+%% @doc Update a `pure_aworset()'.
+-spec mutate(pure_aworset_op(), pure_type:id(), pure_aworset()) ->
+    {ok, pure_aworset()}.
+mutate({add, Elem}, VV, {?TYPE, {POLog, PureAWORSet}}) ->
+    {_, {?TYPE, {POLog0, PureAWORSet0}}} = pure_polog:remove_redundant({VV, {add, Elem}}, {?TYPE, {POLog, PureAWORSet}}),
+    {ok, {?TYPE, {orddict:store(VV, {add, Elem}, POLog0), PureAWORSet0}}};
+mutate({rmv, Elem}, VV, {?TYPE, {POLog, PureAWORSet}}) ->
+    {_, {?TYPE, {POLog0, PureAWORSet0}}} = pure_polog:remove_redundant({VV, {rmv, Elem}}, {?TYPE, {POLog, PureAWORSet}}),
+    {ok, {?TYPE, {POLog0, PureAWORSet0}}}.
+
+%% @doc Returns the value of the `pure_aworset()'.
+%%      This value is a set with all the elements in the `pure_aworset()'.
+-spec query(pure_aworset()) -> sets:set(pure_type:element()).
+query({?TYPE, {POLog0, PureAWORSet0}}) ->
+    Elements0 = ordsets:to_list(PureAWORSet0),
+    Elements1 = [El || {_Key, {_Op, El}} <- orddict:to_list(POLog0)],
+    sets:from_list(lists:usort(lists:append(Elements0, Elements1))).
+
```

Comment:
The sort should be removed here. We were returning and orddict, but it is a
set.

On Tue, May 31, 2016 at 8:13 PM, Carlos Baquero <notifications@github.com>
wrote:

> In src/pure_aworset.erl
> <https://github.com/lasp-lang/types/pull/35#discussion_r65245008>:
>
> > +    {ok, pure_aworset()}.
> > +mutate({add, Elem}, VV, {?TYPE, {POLog, PureAWORSet}}) ->
> > +    {_, {?TYPE, {POLog0, PureAWORSet0}}} = pure_polog:remove_redundant({VV, {add, Elem}}, {?TYPE, {POLog, PureAWORSet}}),
> > +    {ok, {?TYPE, {orddict:store(VV, {add, Elem}, POLog0), PureAWORSet0}}};
> > +mutate({rmv, Elem}, VV, {?TYPE, {POLog, PureAWORSet}}) ->
> > +    {_, {?TYPE, {POLog0, PureAWORSet0}}} = pure_polog:remove_redundant({VV, {rmv, Elem}}, {?TYPE, {POLog, PureAWORSet}}),
> > +    {ok, {?TYPE, {POLog0, PureAWORSet0}}}.
> > +
> > +%% @doc Returns the value of the `pure_aworset()'.
> > +%%      This value is a set with all the elements in the `pure_aworset()'.
> > +-spec query(pure_aworset()) -> sets:set(pure_type:element()).
> > +query({?TYPE, {POLog0, PureAWORSet0}}) ->
> > +    Elements0 = ordsets:to_list(PureAWORSet0),
> > +    Elements1 = [El || {_Key, {_Op, El}} <- orddict:to_list(POLog0)],
> > +    sets:from_list(lists:usort(lists:append(Elements0, Elements1))).
> > +
>
> Might be my ignorance, but is this intermediate list being sorted before
> being transformed into a set?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lasp-lang/types/pull/35/files/9283d4a8f3f2b8533ea90e4ceeba919b92ba120a#r65245008>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ALhjZqbqwH8OJ4rJcqTpNHpTXRRW-_Usks5qHIhOgaJpZM4IqCKz>
> .
>



-- 
Georges Younes
HASLab / INESC TEC &
Universidade do Minho,
Portugal

georges.r.younes@i <cbm@di.uminho.pt>nesctec.pt